File: examples/cpp/example_service_client.cc
//
// Created by pairs on 7/11/18.
//

#include <zros/zros.h>
#include <zros/node_handle.h>
#include "test_service.pb.h"

int main() {
    // 1. init & run service discovery
    // for service discovery
    const std::string master_address = "localhost:23333";
    const std::string agent_address = "localhost:40002"; // listen address for accept register/unregister request
    zros::init(master_address, agent_address);

    // 2. create node
    const std::string node_address = "localhost:50002";
    const std::string node_name = "example_service_client";
    auto node_handle = std::make_shared<zros::NodeHandle>(node_address, node_name);

    // 3. create service client
    const std::string service_name = "test_service";
    auto client = node_handle->serviceClient<zros_example::TestServiceRequest, zros_example::TestServiceResponse>(
            service_name
            );

    std::thread call_thread = std::thread([client]() {
        while (true) {
            zros_example::TestServiceRequest request;
            zros_example::TestServiceResponse response;
            while(client->get_ready()) {
                auto status = client->call(&request, &response);
                if (status.code() == status.OK) {
                    SSPD_LOG_INFO << "call success";
//                    SSPD_LOG_INFO << "the response is " << response;
                } else {
                    SSPD_LOG_WARNING << "call failed " << status.details();
                }
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
    });
    call_thread.join();
}

File: examples/python/example_service_client.py
# -*- coding: utf-8 -*-
""" An example for service client

data: 2018-08-01
author: pairs
"""

import zros_python.zros as zros
import zros_python.node_handle as node
import zros_python.pb.test_service_pb2 as tspb2
import time
import traceback
import logging

logger = logging.getLogger(__name__)

_TICK_TASK_TIME = 1

def call():
    # 1. init & run service discovery
    master_address = u'localhost:23333'
    zros.init(master_address)

    # 2. create node
    node_name = u'example_service_client'
    node_handle = node.NodeHandle(node_name)

    # 3. create service client
    service_name = u'test_service'
    client = node_handle.service_client(service_name,
                               tspb2.TestServiceRequest,
                               tspb2.TestServiceResponse)
    if client is None:
        logger.error(u'service client %s failed', service_name)
        return
    # 4. call
    request = tspb2.TestServiceRequest()
    request.detail = u'hello world'
    while True:
        try:
            if client.get_ready() is True:
                res = client.call(request)
                if res is None:
                    logger.warning(u'call service %s failed', service_name)
                else:
                    logger.info(u'call service %s success', service_name)
                    logger.info(u'response detail is %s ', res.detail)
        except Exception:
            logger.error(traceback.format_exc())
        time.sleep(_TICK_TASK_TIME)

if __name__ == '__main__':
    logging.basicConfig(format=u'[%(name)s-%(levelname)s]: %(message)s',
                        level=logging.NOTSET)
    call()